brick-editor.js

```diff
@@ -35,163 +37,487 @@ function addBlocksHTML() { // eslint-disable-line no-unused-vars
+                var response = confirm("Are you sure you wish to delete?");
+                if (response) {
+                    editor.setValue(deleteBlock(ast, node));
+                } else {
+                    unhighlight();
+                }
+            }, 100);
+        } else {
+            // delete char
+            editor.setValue(deleteChar(buffer, cursor));
+            setCursor(cursor);
+        }
+    }
+}
+
+/**
+ * Handle button clicks
  *
  * @param {number} i - Index of code in dictionary
  * @returns {undefined}
  */
 function buttonHandler(i) { // eslint-disable-line no-unused-vars
-    var template = blockDict[i]["code"];
+    var template = blockDict[i].code;
     var ast = recast.parse(editor.getValue());
-    var position = getPosition();
+    var cursor = getCursor();
 
     // add block to buffer string and update editor
-    var new_text = addBlock(template, ast, position);
+    var new_text = addBlock(template, ast, cursor);
     ast = recast.parse(new_text);
     editor.setValue(recast.print(ast).code);
-   
+
     // update cursor position
-    // FIXME calculate new position
-    setPosition(position);
+    // FIXME calculate new cursor
+    setCursor(cursor);
 }
 
 // EDITOR INTERFACE CODE
 
 /**
- * Get the cursor position in the editor.
+ * Make a Cursor object.
+ *
+ * @param {int} line - The line number.
+ * @param {int} col - The column number.
+ * @returns {Cursor} - The Cursor object.
+ */
+function makeCursor(line, col) {
+    return { "lineNumber": line, "column": col };
+}
+
+/**
+ * Get the cursor cursor in the editor.
  *
  * Note monaco line number starts at 0, while recast line number starts at 1.
  * The result of this function, and all functions here, follow the recast
  * numbering.
  *
- * @returns {Location} The line and column number of the cursor
+ * @returns {Cursor} - The line and column of the cursor in the editor.
  */
-function getPosition() {
-    var position = editor.getPosition();
-    position.column = position.column - 1;
-    return position;
+function getCursor() {
+    var cursor = editor.getPosition();
+    cursor.column = cursor.column - 1;
+    return cursor;
 }
 
 /**
- * Get the cursor position in the editor.
+ * Get the cursor cursor in the editor.
  *
  * Note monaco line number starts at 0, while recast line number starts at 1.
  * The input of this function follows the recast numbering.
  *
- * @param {Location} position - A LineNumber and Column object.
+ * @param {Cursor} cursor - The line and column of the cursor.
+ * @returns {undefined}
+ */
+function setCursor(cursor) {
+    cursor.column = cursor.column + 1;
+    editor.setPosition(cursor);
+}
+
+/**
+ * Get the positions of the selection in the editor
+ * (Offsets the start and end columns by 1 to account for differences in column numbering between Recast and Monaco)
+ *
+ * @returns {Location} An object with start/end lineNumber and start/end column properties
+ */
+function getSelection() {
```

Comment:
Is the lineNumber going to be problematic here? If so, it might be worth making another set of adaptor functions.